tests/phpunit/includes/bootstrap.php

```diff
@@ -76,8 +76,13 @@
 		if ( is_string( WP_TESTS_PHPUNIT_POLYFILLS_PATH )
 			&& '' !== WP_TESTS_PHPUNIT_POLYFILLS_PATH
 		) {
-			$phpunit_polyfills_path = rtrim( $phpunit_polyfills_path, '/\\' );
-			$phpunit_polyfills_path = realpath( $phpunit_polyfills_path . '/phpunitpolyfills-autoload.php' );
+			// Be tolerant to the path being provided including the filename.
+			if ( substr( $phpunit_polyfills_path, -29 ) !== 'phpunitpolyfills-autoload.php' ) {
+				$phpunit_polyfills_path = rtrim( $phpunit_polyfills_path, '/\\' );
+				$phpunit_polyfills_path = $phpunit_polyfills_path . '/phpunitpolyfills-autoload.php';
+			}
+
+			$phpunit_polyfills_path = realpath( $phpunit_polyfills_path );
```

Comment:
I don't think that there's any actual benefit to be had from using `realpath()` here, but it adds two additional liabilities to the code:
- it does not work on virtualized filesystems, which might be a problem on managed hosting as they might use a virtualized FS to put files on NAS.
- it does not work when one or more of the portions that make up the path are not executable to the script, which might be a problem on shared hosting, which might give access to home folders, but not to the entire FS tree.

I know that this is more of a development use case and might not often be run on production servers, but on strictly tested sites, the tests will run on the actual machine that a deployment is targeting, so it could be a potential issue.

Given that the path is only used for a single `require_once` in the same file, it is safer to just use the provided path as-is without trying to normalize it (and potentially fail on the normalization).

In addition, that makes sure that the error message properly reflects what what executed. Otherwise, the `realpath()` might come up with something completely different, fail for some reason, and then the error message shows the path before the `realpath()`, making it hard to debug. The opposite would of course be just as confusing, showing the `realpath()` in the error message, which might be not at all what the user provided (due to symlinks, etc...).

```suggestion
```

Reply:
@schlessera I understand where you are coming from and agree this can be further improved, but if `realpath()` is not used, we need to make it explicit whether the constant should be set with an absolute or a relative path and if a relative path, it should be made explicit what the path is relative to.

An absolute path is generally a bad idea as it doesn't transfer between systems, unless it is set dynamically in the bootstrap file calling this bootstrap file, i.e. it would break the option of setting the path via the `phpunit.xml.dist` file.

A relative path has its own drawbacks as the "relative to" needs to be clearly defined in that case.

So, I suggest we go with a relative path, but would like to see some input of what this path should be relative to.

Some options:
* Relative to the WP Core `bootstrap.php` file.
* Relative to the WP Core `tests` directory.
* Relative to the current working directory.
* Relative to the WP root directory - note: this option would not allow for the `scaffold-command` setup.

Opinions ? Other options ?

Reply:
To be honest, I didn't even consider a relative path here in the constant, as I thought of it as an environment variable more so than a site-specific constant.

Why not just require an absolute path here and avoid the ambiguity of a relative path? The absolute path can easily be set using any of the other constants that let you smoothly transfer the files, like `ABSPATH`, `WP_INC`, or even `HOME`.

Any relative scenario will be very confusing, as you don't know _where_ a developer will define that constant.

Reply:
Okay, so we have a conundrum here as using absolute paths will basically make it impossible to set this via the `phpunit.xml.dist` file, while that is typically the location for setting such constants.

Let's ask some more people for opinions.
/cc @swissspidy @jeherve @thomasplevy @hellofromtonya

Reply:
Both of you have valid points for the `WP_TESTS_PHPUNIT_POLYFILLS_PATH ` constant's value:

- relative path is needed when defined in `phpunit.xml.dist` 
- else, absolute path to let the developer define the exact location

How can we solve this for both scenarios? Maybe separate constants (where usage of how to define is clearly specified in the dev note)?

Reply:
A plugin/theme setting up integration tests needs a `bootstrap.php` anyway to load and configure the WP test suite, so it to me it sounds reasonable to have to define such a constant in that file to make the integration tests work. I wouldn't even have considered putting that into the XML config file. Agreed that using relative paths there would be confusing, I wouldn't have considered using that either.

Just my 2 cents :)

Reply:
Jumping in with my 2 cents as well :) I would also vote for an absolute path set in the bootstrap.php file too. I believe that's a relatively easy way of setting things up in plugins.

Reply:
Please note that the constant is not only for integrators, but also for select Core developers who prefer not to use Composer as described in this workflow in the handbook: https://make.wordpress.org/core/handbook/testing/automated-testing/phpunit/#setting-up-to-run-with-the-phpunit-phar-file-without-composer (yes, we updated that page recently).
In that case, there will be no `bootstrap.php` file in which the constant can be set.

Reply:
In those cases, could the constant be defined in `wp-tests-config.php`, or is `phpunit.xml.dist` the only option then?

Reply:
It could be, but the `phpunit.xml` (overload) file is the typical location for these type of constants, similar to how the `WP_RUN_CORE_TESTS` and `WP_TESTS_MULTISITE` get set, so, while setting the constant in `wp-test-config.php` is perfectly possible, it is counter-intuitive and inconsistent with how other constants used in the core test bootstrap are set.

Reply:
Okay, so to get this PR out of this stalemate, I've now updated it to expect an absolute path in the constants.

For **WP Core developers** who want to use the ["PHPUnit PHAR file without Composer"](https://make.wordpress.org/core/handbook/testing/automated-testing/phpunit/#setting-up-to-run-with-the-phpunit-phar-file-without-composer) workflow, this will still allow them to set the constant in the `phpunit.xml` overload file as the overload file is specific to their system, so setting an absolute path will not be problematic.

For **plugin/theme integration tests**, this means the constant must be set in the test bootstrap of the plugin/theme itself **before** loading the WP native test `bootstrap.php` file.

I've updated the inline documentation and the error messages to be in line with these changes.

Reply:
P.S.: review and testing of this latest iteration would be highly appreciated!!!